binaries/oli/src/config/mod.rs

```diff
@@ -0,0 +1,159 @@
+        Config::load_from_str(&data)
+    }
+
+    pub(crate) fn load_from_str(s: &str) -> Result<Config> {
+        Ok(toml::from_str(s)?)
+    }
+
+    /// Parse `<profile>://abc/def` into `op` and `location`.
+    pub fn parse_location<'a>(&self, s: &'a str) -> Result<(Operator, &'a str)> {
+        if !s.contains("://") {
+            let mut fs = services::Fs::default();
+
+            let filename = match s.rsplit_once(['/', '\\']) {
+                Some((base, filename)) => {
+                    fs.root(base);
+                    filename
+                }
+                None => s,
+            };
+
+            return Ok((Operator::new(fs)?.finish(), filename));
+        }
+
+        let parts = s.splitn(2, "://").collect::<Vec<_>>();
+        debug_assert!(parts.len() == 2);
+
+        let profile_name = parts[0];
+
+        let bucket_and_path = parts[1].splitn(2, '/').collect::<Vec<_>>();
+        debug_assert!(bucket_and_path.len() == 2);
+
+        let bucket = bucket_and_path[0];
+        let path = bucket_and_path[1];
+
+        let profile = self
+            .profiles
+            .get(profile_name)
+            .ok_or_else(|| anyhow!("unknown profile: {}", profile_name))?;
+
+        if profile.s3.is_some() {
+            let opts = set_bucket!(profile.s3, bucket);
+            Ok((Operator::from_map::<services::S3>(opts)?.finish(), path))
+        } else if profile.oss.is_some() {
+            let opts = set_bucket!(profile.oss, bucket);
+            Ok((Operator::from_map::<services::Oss>(opts)?.finish(), path))
+        } else {
+            Err(anyhow!("invalid profile"))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use opendal::Scheme;
+
+    #[test]
+    fn test_load_toml() {
+        let cfg = Config::load_from_str(
+            r#"
+[profiles.mys3.s3]
+region = "us-east-1"
+access_key_id = "foo"
+enable_virtual_host_style = "on"
+"#,
+        )
+        .expect("load config");
+        let profile = cfg.profiles["mys3"].s3.clone().unwrap();
+        assert_eq!(profile["region"], "us-east-1");
+        assert_eq!(profile["access_key_id"], "foo");
+        assert_eq!(profile["enable_virtual_host_style"], "on");
+    }
+
+    #[test]
+    fn test_parse_fs_location() {
+        let cfg = Config::default();
+        let (op, path) = cfg.parse_location("./foo/bar/1.txt").unwrap();
+        assert_eq!("1.txt", path);
+        let info = op.info();
+        assert!(info.root().ends_with("foo/bar"));
+        assert_eq!(Scheme::Fs, info.scheme());
+    }
+
+    #[test]
+    fn test_parse_s3_location() {
+        let cfg = Config {
+            profiles: StringMap::from([(
+                "mys3".into(),
+                Profile {
+                    s3: Some(StringMap::from([("region".into(), "us-east-1".into())])),
+                    ..Default::default()
+                },
+            )]),
+        };
+        let (op, path) = cfg.parse_location("mys3://mybucket/foo/1.txt").unwrap();
```

Comment:
Maybe we can treat `bucket` as a config value?

Reply:
I feel it's natural that 1 profile corresponds to a bucket (an OpenDAL operator), but also understand the need to access different buckets without duplicating buckets. 

Maybe we need something like

```toml
[s3]
region = "us-east-1"
```

which is a default configuration for `s3` (can be overriden by profiles). Then we use it like
```
oli cp s3://... s3://...
```
instead of profiles.

Reply:
I found another fact: actually 
```
oli cp s3://bucket-a/.. s3://bucket-b/...
```
is already supported (before this PR). You can try it.

Because of this I think including bucket in profile makes more sense.

Reply:
Please correct me if I misunderstand what your proposal is:
You mean we should save bucket in the profile, so that a profile 1:1 maps to an OpenDAL operator. Besides, OpenDAL is already able to read credentials from aws cli config, we don't need to save credentials in the profile. Therefore, it is convenient to omit bucket from the uri.

After deliberation, I think it makes sense. So the usage of `oli cp` should be
```
oli cp profile1://<path> profile2://<path>
```
right?

Reply:
Done. PTAL

Reply:
> You mean we should save bucket in the profile, so that a profile 1:1 maps to an OpenDAL operator. 

This is exactly my idea.

> Besides, OpenDAL is already able to read credentials from aws cli config, we don't need to save credentials in the profile.

This looks also good. 🤔 But actually I mean we treat `s3://` specially. i.e., it's a **scheme** instead of a **profile**, which is the previous implementation (but you didn't delete it which confused me a little..). And in this case, we use the standard s3uri to parse it, and the UX for `oli` is similar to `aws s3`. We can access any bucket without defining a profile first.

But maybe we can KISS now in the MVP and just include the profile design.

Reply:
> I mean we treat s3:// specially. i.e., it's a scheme instead of a profile,

What if a user need to copy data across different accounts, say `oli cp s3://bucket1/<path> s3://bucket2-in-another-account/<path>` 🤔.

> We can access any bucket without defining a profile first.

If you are worrying about the requirement of a profile before running `oli`, I think we could treat the `s3` scheme as a special profile. If a uri like `s3://bucket/<path>` is supplied, then we could use a default S3Builder.

> but you didn't delete it which confused me a little..

Thanks for reminding me of this, I actually forgot 😂

Reply:
> What if a user need to copy data across different accounts

Please don't over thinking use cases before we really meet it. I prefer to deliver the minimum workable product. Let's focus on the simplest case.